File: XJTU-COMP461205-CN-main/软件实验/Scoket编程实验/TCP/TCP-Client.cpp
#include <iostream>
#include <string>
#include <thread>
#include <atomic>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <string.h>

std::atomic<bool> exit_flag(false);
std::atomic<bool> proof_flag(false);
std::atomic<bool> recv_flag(true);
int client_socket;

void wait_for_exit() 
{
    while (!exit_flag.load()) {}
    close(client_socket);
    exit(0);
}

void CS_Server() 
{
    char buffer[1024];
    while (true) 
    {
        if (exit_flag.load()) return;

        int received = recv(client_socket, buffer, 1024, 0);
        if (received <= 0) continue;

        buffer[received] = '\0'; // 确保字符串以空字符结束
        std::string data(buffer);

        if (proof_flag.load()) 
        {
            std::cout << "服务器: " << data << std::endl;
            recv_flag.store(true);
        } 
        else 
        {
            std::cout << data;
            if (data == "信息正确!\r\n") 
            {
                proof_flag.store(true);
                std::cout << "客户端: ";
            }
        }
    }
}

void CS_Client() 
{
    while (true) 
    {
        std::string data;
        if (proof_flag.load()) 
        {
            while (!recv_flag.load()) {}
            std::cout << "客户端: ";
            std::getline(std::cin, data);
            recv_flag.store(false);
        } 
        else 
        {
            std::getline(std::cin, data);
            recv_flag.store(false);
        }

        send(client_socket, data.c_str(), data.length(), 0);
        if (data == "exit") 
        {
            exit_flag.store(true);
            return;
        }
    }
}

int main() {
    client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket < 0) 
    {
        std::cerr << "Socket creation failed." << std::endl;
        return -1;
    }

    sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(1212);
    inet_pton(AF_INET, "127.0.0.1", &server_addr.sin_addr); // 使用 localhost

    if (connect(client_socket, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) 
    {
        std::cerr << "Connection failed." << std::endl;
        return -1;
    }

    std::cout << "Connect Succeed!" << std::endl;

    std::thread exit_thread(wait_for_exit);
    std::thread server_thread(CS_Server);
    std::thread client_thread(CS_Client);

    exit_thread.detach();
    server_thread.detach();
    client_thread.join();

    return 0;
}

File: XJTU-COMP461205-CN-main/软件实验/Scoket编程实验/TCP/TCP-Client.py
import threading
import socket
import sys
import os

# data
exit_flag = False
proof_flag = False
recv_flag = True

def wait_for_exit(sock):
    global exit_flag, proof_flag
    while True:
        if exit_flag:
            sock.close()
            os._exit(0)
            break
    return

def CS_Server(sock):
    global exit_flag, proof_flag, recv_flag
    proof_flag = False
    while True:
        try:
            if exit_flag:
                return
            data = sock.recv(1024)
            data = data.decode('utf8')
            if(proof_flag):
                print('服务器:', data)
                sys.stdout.flush()
                recv_flag = True
            else:
                print(data, end='')
                sys.stdout.flush()
            if(data == '信息正确!\r\n'):
                proof_flag = True
                print('客户端: ', end='')
                sys.stdout.flush()
        except:
            pass    

def CS_Client(sock):
    global exit_flag, proof_flag, recv_flag
    while True:
        try:
            if(proof_flag):
                while(recv_flag == False):
                    pass
                data = input('客户端: ')
                recv_flag = False
            else:
                data = input('')
                recv_flag = False
        except:
            continue
        sock.send(data.encode('utf8'))
        if(data == 'exit'):
            exit_flag = True
            return
        

# main
client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
host = socket.gethostname()
port = 1212
client_socket.connect((host, port))
print("Connect Succeed!")
sys.stdout.flush()
threading.Thread(target=wait_for_exit, args=(client_socket,)).start()
threading.Thread(target=CS_Server, args=(client_socket,)).start()
threading.Thread(target=CS_Client, args=(client_socket,)).start()
